### ISSUE TITLE
Suggestion: consider ternary statements that select between two constant expressions to be leaf nodes

### ISSUE THREAD

**Issue by BernieSumption** [istanbuljs/istanbuljs#221]
I have a feature suggestion and would be interested in feedback. The title of this issue says it all, here's my justification.

When a ternary is used to select between two constant expressions, executing the ternary once is enough to guarantee that both potential values would work.

```js
const k = true;
const v = k ? "ya" : "nay";
```

If this statement is executed once, and `k` is true, the only difference you will get by executing it again with `k` being false would be to change the value of `v`. As such, this is the only instance in coverage I can think of where istanbul requires full coverage of all possible *values*.

I'd suggest that it makes sense for istanbul to consider the above ternary to be a leaf expression, rather than a node with two branches.

If the ternary contains any expressions, even a simple variable lookup, I'd call it a branch, e.g. `k ? a : b` should be considered a node with two branches because a or b could potentially be unbound giving a ReferenceError.

**Comment by BernieSumption** [comment]
Note: there is a workaround that I use for projects that use a lot of constant lookup ternaries:

```
function select(test, ifTrue, ifFalse) {
    return test ? ifTrue : ifFalse;
}
```

By using `select(k, "ya", "nay")` in place of `k ? "ya" : "nay"`, and ensuring that the `select` function has a test with both values, I can avoid having to write two test cases for every constant selection in the codebase.

But it would be nice to get this behaviour with idiomatic JS.

**Comment by coreyfarrell** [comment]
I disagree with your assessment.  It's not just about the success/failure to run the statements in true/false branches of the tertiary, those different values could be used later in code and it's possible that one of the values could expose a bug in other parts of your code which use the value of `v`.  This includes situations where maybe `v == "nay"` is off-nominal, you might have later code which fails to deal with the off-nominal value.

If you really want to suppress coverage for a tertiary you could try one of the parsing hints.  See https://github.com/istanbuljs/nyc#parsing-hints-ignoring-lines

**Comment by BernieSumption** [comment]
If anyone wants to reject this request on the grounds that it's a silly edge case, not required by most users, not worth the complexity/value trade-off in implementation or whatever then totally fair.

But your argument that a ternary producing a different value could lead to a bug applies to any statement, such as indexing an array or even multiplying two numbers. If a statement is given different operands then it will evaluate to a different value. And this, could in principle, cause a bug. But ternaries are the only place where this causes less than 100% code coverage.

I don't see why ternaries selecting between constant expressions should be a unique special case where 100% value coverage is required.

**Comment by stereobooster** [comment]
I think this is intersects with different problem - unreachable code. There are other tools that are in better position to detect unreachable code, like [TypeScript](https://www.typescriptlang.org/tsconfig/#allowUnreachableCode)
